### ISSUE TITLE
diag(p3): BC-init specialist then continue PPO — basin-trap vs anti-attractor

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#270]
## Hypothesis being tested

After 5 plausible PPO-side interventions all returned tier-3 verdicts (#225/#231 MAPPO, #220/#233 obs-fix, #221/#228 spatial, #260/#266 curriculum, #259/#262/#268 action-shaping), we don't yet know whether:

- **(A) Basin trap**: the specialist's policy region is stable under PPO updates but unreachable from random init
- **(B) Anti-attractor**: something in PPO's update rule actively pulls policies AWAY from specialist behavior

This experiment cleanly distinguishes A from B in a single training run.

## Protocol

1. **BC pretrain**: gather ~10k specialist (heuristic) demonstrations on `minimal_specialization`. Behaviorally clone the policy network for ~1k SGD steps until BC loss converges. Verify the resulting policy approximates the specialist (gap_closed ≥ 0.7 on eval before PPO starts).
2. **Continue PPO**: run standard IPPO for 50 iterations from the BC-init checkpoint. 3 seeds.
3. **Compare trajectories** against random-init PPO (we have these from PR #257).

## Verdict logic

| BC-init + PPO outcome | Interpretation | Next intervention |
|---|---|---|
| Reward holds at near-specialist (gap_closed > 0.5 throughout) | **Basin trap confirmed.** Specialist is stable under PPO; just unreachable from random init. | PBT, neuroevolution, RND/exploration; or BC-init as production technique |
| Reward decays back toward random | **Anti-attractor confirmed.** Specialist isn't even a local attractor for PPO. Something in our reward/dynamics actively repels specialist-like behavior. | Examine WHY — likely the reward structure has a counterintuitive gradient near specialist |
| Reward stays elevated but below specialist | Partial basin / local minimum near specialist | Hybrid: combine BC-init with mild reward shaping (#259/#265) |

## Cost

~3 hours wall-clock on `COMPUTE_HOST_PRIMARY`. Cheap relative to information value.

## Acceptance criteria

- [ ] BC pretraining script in `experiments/p3_specialization/bc_init.py`
- [ ] Verified BC policy approximates specialist on eval before PPO
- [ ] 50-iter PPO continuation, 3 seeds
- [ ] Notebook entry with verdict + interpretation
- [ ] Project memory updated based on outcome

## References

- PR #268 (action-shaping tier-3 verdict, motivates this)
- PR #257 (random-init PPO baseline for comparison)
- docs/ppo-help-brief.md (broader context)

**Comment by rjwalters** [comment]
## Curator Enhancement (2026-05-17)

### Codebase Context

Confirmed the implementation primitives exist and are stable. Pointers below are file:line-anchored against `main` at curation time.

**Specialist API (already callable, no work needed):**
- `bucket_brigade/baselines/specialist.py:119` — `specialist_action_joint(obs, num_agents, num_houses=10) -> (N, 3) int64`. Pure-functional, zero state, output shape matches `MultiDiscrete([10, 2, 2])` post-#235 (honest-signaling: `signal == mode`).
- `bucket_brigade/baselines/specialist.py:139` — `SpecialistPolicy(num_agents).__call__(obs)` callable wrapper, identical shape contract.
- Demonstration generation pattern lives at `experiments/p3_specialization/diagnostics/issue199_baselines.py:68-76` — episode rollout calling `specialist_action_joint`. Copy this pattern; on `minimal_specialization` an episode is ~5-200 nights, so 10k (obs, action) pairs is ~50-200 episodes (<1 minute CPU).

**Policy network checkpoint loading (already supported):**
- `bucket_brigade/training/networks.py:13` — `PolicyNetwork(obs_dim, action_dims, hidden_size)` is a standard `nn.Module` with state_dict save/load. The 3 action heads (`action_heads[0/1/2]` for house/mode/signal) match `action_dims=[10, 2, 2]`.
- `bucket_brigade/training/joint_trainer.py:259-268` — `self.policies = nn.ModuleList([PolicyNetwork(...) for _ in range(num_agents)])`. Per-agent state-dicts are saved at `experiments/p3_specialization/train.py:719-720` via `torch.save(policy.state_dict(), pol_dir / f"agent_{i}.pt")` — that's the load target.
- **`obs_dim` includes the per-agent identity one-hot tail** (issue #204). For `num_agents=4, num_houses=10`: obs_dim = 42 + 4 = 46. The BC trainer must flatten observations with `flatten_dict_obs(obs, agent_id=i, num_agents=4)` (`bucket_brigade/training/joint_trainer.py:62`) to match — feeding obs without the identity tail to a `PolicyNetwork` built for obs_dim=46 is a silent dimension mismatch.

**Train script injection point:**
- `experiments/p3_specialization/train.py:606` — `train_one_cell(cfg, output_dir)` constructs `JointPPOTrainer` at line 634 then drives the iter loop. Add `--bc-init` CLI flag + `bc_init_checkpoint_dir: Optional[Path]` field on `CellConfig` (line 56). After the trainer is constructed, load per-agent state dicts: `trainer.policies[i].load_state_dict(torch.load(bc_init_dir / f"agent_{i}.pt"))` before the iteration loop. Centralized critic (line 280-298) is independent and uninitialized by BC — that's fine, only enable IPPO for this experiment.

**Eval / verdict tooling exists:**
- `experiments/p3_specialization/analyze_220.py:57-60` — `MINSPEC_RANDOM = -96.07`, `MINSPEC_SPECIALIST = -22.07`, `gap_closed(ppo_trailing5) = (ppo - random) / (specialist - random)`. Reuse this constants module directly in BC-fit eval and PPO continuation analysis.

**No existing BC infrastructure** — `grep -r behavioral_clon\|imitation bucket_brigade/training` returns zero hits. This issue creates the BC pipeline from scratch.

### Implementation Sketch

**New file: `experiments/p3_specialization/bc_init.py`** (~150 LOC):

```python
"""Behaviorally clone the specialist policy, then save per-agent checkpoints
compatible with JointPPOTrainer.policies[i].load_state_dict()."""
import argparse, torch, numpy as np
from pathlib import Path
from torch.utils.data import DataLoader, TensorDataset
from bucket_brigade.baselines import specialist_action_joint
from bucket_brigade.envs.bucket_brigade_env import BucketBrigadeEnv
from bucket_brigade.envs.scenarios_generated import get_scenario_by_name
from bucket_brigade.training.networks import PolicyNetwork
from bucket_brigade.training.joint_trainer import flatten_dict_obs

def gather_demos(scenario_name, num_agents, num_pairs, seed):
    """Roll specialist out until we have num_pairs (per-agent-obs, action) tuples."""
    scenario = get_scenario_by_name(scenario_name, num_agents=num_agents)
    env = BucketBrigadeEnv(scenario=scenario)
    rng = np.random.default_rng(seed)
    obs_buf = [[] for _ in range(num_agents)]   # per-agent flat obs
    act_buf = [[] for _ in range(num_agents)]   # per-agent (3,) actions
    n = 0
    while n < num_pairs:
        obs = env.reset(seed=int(rng.integers(0, 2**31 - 1)))
        while not env.done:
            joint = specialist_action_joint(obs, num_agents, num_houses=10)  # (N, 3)
            for i in range(num_agents):
                obs_buf[i].append(flatten_dict_obs(obs, agent_id=i, num_agents=num_agents))
                act_buf[i].append(joint[i])
            obs, _, _, _ = env.step(joint)
            n += 1
            if n >= num_pairs: break
    return [np.array(o) for o in obs_buf], [np.array(a) for a in act_buf]

def bc_fit(obs, acts, obs_dim, action_dims, epochs, batch_size, lr, device):
    """Cross-entropy supervised fit on each of the 3 action heads, summed."""
    policy = PolicyNetwork(obs_dim=obs_dim, action_dims=action_dims, hidden_size=64).to(device)
    opt = torch.optim.Adam(policy.parameters(), lr=lr)
    ds = TensorDataset(torch.from_numpy(obs).float(), torch.from_numpy(acts).long())
    dl = DataLoader(ds, batch_size=batch_size, shuffle=True)
    for ep in range(epochs):
        for x, a in dl:
            x, a = x.to(device), a.to(device)
            logits, _ = policy(x)  # list of 3 tensors
            loss = sum(torch.nn.functional.cross_entropy(logits[k], a[:, k]) for k in range(3))
            opt.zero_grad(); loss.backward(); opt.step()
    return policy

# Main: gather demos, fit each of N agents' policies (each fits on its own per-agent
# observation stream — identity-tail differs per agent), save state_dicts to
# output_dir/agent_{i}.pt. Then run eval episode against env and report gap_closed.
```

**Modify `experiments/p3_specialization/train.py`:**

Add to `CellConfig` (after line 111):
```python
    # Issue #270: optional BC-pretrained init. Directory containing
    # agent_{i}.pt state dicts produced by bc_init.py. None = random init
    # (default, preserves bit-identical pre-#270 behavior).
    bc_init_checkpoint_dir: Optional[str] = None
```

After trainer construction (around line 650), before the iteration loop:
```python
    if cfg.bc_init_checkpoint_dir is not None:
        bc_dir = Path(cfg.bc_init_checkpoint_dir)
        for i, policy in enumerate(trainer.policies):
            sd = torch.load(bc_dir / f"agent_{i}.pt", map_location=cfg.device)
            policy.load_state_dict(sd)
        print(f"  BC-init: loaded {cfg.num_agents} policies from {bc_dir}")
```

Add CLI flag (in `main()` around line 813):
```python
    p.add_argument("--bc-init-checkpoint-dir", type=str, default=None,
                   help="Issue #270: load BC-pretrained per-agent state dicts before PPO.")
```

### Hyperparameter Starting Points (BC phase)

| Hparam | Starting value | Rationale |
|---|---|---|
| Demo count | 10000 (obs, action) pairs **per agent** = 40k total across N=4 | Issue brief calls for 10k; multiply by N because each agent has its own identity one-hot |
| BC epochs | 10 | Specialist policy is a 2-rule deterministic mapping; convergence should be fast |
| Batch size | 256 | Matches `JointPPOTrainer.minibatch_size` default |
| Learning rate | 1e-3 | Higher than PPO's 3e-4 — supervised loss is well-conditioned, no clip range |
| Optimizer | Adam | Match PPO for fair continuation |
| Loss | Sum of CE on each of 3 action heads | Independent factorization matches PolicyNetwork output |

### Verdict-Gating Eval (between BC and PPO)

Per the issue's protocol step 1, **must verify gap_closed ≥ 0.7 before launching the PPO continuation** — otherwise the BC didn't take and the experiment is uninterpretable. Eval pattern:

```python
# After BC, before saving — run eval episodes
team_rewards = []
for ep in range(20):
    obs = env.reset(seed=ep)
    ep_reward = 0.0
    while not env.done:
        # Stack per-agent observations
        joint = []
        for i in range(num_agents):
            x = flatten_dict_obs(obs, agent_id=i, num_agents=num_agents)
            actions, _, _ = policies[i].get_action(torch.from_numpy(x).float().unsqueeze(0), deterministic=True)
            joint.append([a.item() for a in actions])
        obs, rewards, _, _ = env.step(np.array(joint, dtype=np.int64))
        ep_reward += rewards.sum()
    team_rewards.append(ep_reward)
mean_step = np.mean(team_rewards) / mean_episode_length
gc = (mean_step - MINSPEC_RANDOM) / (MINSPEC_SPECIALIST - MINSPEC_RANDOM)
assert gc >= 0.7, f"BC didn't take: gap_closed={gc:.3f}, abort PPO continuation"
```

If gap_closed < 0.7, this likely points to a representational gap in `PolicyNetwork(hidden_size=64)` — which is exactly what sister issue **#272 ("specialist BC-fit alone")** is designed to characterize first. **Strongly recommend completing #272 before this issue** to surface architecture issues cheaply. See Dependencies section.

### Wall-Clock Budget (~3 hours)

| Phase | Time | Where to run |
|---|---|---|
| BC demo gathering (40k pairs) | ~2 min | local OK (specialist is hand-coded, CPU-only) |
| BC supervised fit (10 epochs × N=4 agents) | ~5 min | local OK |
| BC eval (20 episodes) + gap_closed check | ~1 min | local OK |
| **PPO continuation (50 iters × 3 seeds)** | **~2.5 hours** | **`COMPUTE_HOST_PRIMARY`** per CLAUDE.md (each PPO cell ~50 min wall-clock) |
| Analysis + notebook entry | ~10 min | local |

Total fits comfortably in the "~3 hours wall-clock" estimate from the issue body.

### tmux Session Naming

Per CLAUDE.md remote-execution conventions:
- `tmux new -s bc_init_270` — BC pretraining (local or remote)
- `tmux new -s ppo_continue_270_seed_42|43|44` — three PPO seeds in parallel on `COMPUTE_HOST_PRIMARY`

### Acceptance Criteria (refined from issue body)

- [ ] `experiments/p3_specialization/bc_init.py` — demo gen + BC fit + per-agent checkpoint save
- [ ] BC eval reports gap_closed ≥ 0.7 (else: open architecture-fix follow-up, do NOT proceed)
- [ ] CLI flag `--bc-init-checkpoint-dir` plumbed in `train.py` with `bc_init_checkpoint_dir` field on `CellConfig`
- [ ] 50-iter PPO continuation, 3 seeds (42/43/44), `lambda_red=0`, IPPO (no centralized critic, no λ_red)
- [ ] Comparison plot: BC-init PPO trajectory vs random-init PPO baseline (cells under `runs/issue220_treatment/minimal_specialization/lambda_0e0/`, seeds 42/43/44)
- [ ] Research-notebook entry with verdict matrix from issue body (basin trap / anti-attractor / partial)
- [ ] Project memory `project_ppo_failure_mode.md` updated based on outcome

### Affected Files

- `experiments/p3_specialization/bc_init.py` (NEW) — BC demo gen + fit + checkpoint save + eval
- `experiments/p3_specialization/train.py` — add `--bc-init-checkpoint-dir` CLI flag + `bc_init_checkpoint_dir` config field + load loop after trainer construction (line ~650)
- `experiments/p3_specialization/analyze_270.py` (NEW) — compare BC-init vs random-init PPO trajectories; produce verdict per the 3-way matrix in the issue body
- `research_notebook/2026-05-XX_issue270_bc_init_continuation.md` (NEW) — notebook entry

### Test Plan

- [ ] Manual verification: `bc_init.py` runs end-to-end and produces 4 `agent_{i}.pt` files; loading them into a fresh `PolicyNetwork` succeeds with no shape mismatch
- [ ] Manual verification: BC eval prints `gap_closed=0.XX` to stdout, blocks PPO continuation if < 0.7
- [ ] Manual verification: `train.py --bc-init-checkpoint-dir <bc_out>` loads cleanly and the first iteration's `team_reward` is in the specialist-like range (≈ −22, not ≈ −96)
- [ ] Automated test: `tests/test_bc_init.py` — round-trip (gather 100 demos, fit 1 epoch, load into `JointPPOTrainer`, verify state dict equality after load)
- [ ] Edge case: empty checkpoint dir, missing `agent_2.pt`, wrong-shape state dict (each should raise a clear error, not silently no-op)
- [ ] Verify post-load behavior — first PPO rollout's mean step reward should land near specialist (not random), confirming the load worked

### Dependencies

- [ ] **#272** — "specialist BC-fit alone — can the policy network even represent the specialist?" *Recommended completion before this issue.* If #272 finds an architectural representational gap (BC accuracy < 90%), the basin-trap/anti-attractor distinction this issue draws is meaningless until the architecture is fixed. #272 is a 30-min check; it's worth running first.

### References

- Sister issue #272 — specialist BC-fit alone (representational pre-check)
- Sister issue #271 — random-init MLP best-of-N (lottery-ticket pre-check)
- PR #257 — was incorrectly cited in the issue body as "random-init PPO baseline"; the actual random-init `minimal_specialization` baseline cells live at `experiments/p3_specialization/runs/issue220_treatment/minimal_specialization/lambda_0e0/seed_{42,43,44}` (post-#216 obs-fix arm, treatment = random-init IPPO)
- `docs/ppo-help-brief.md` — broader research context
- `MEMORY.md` — PPO-tuning ladder (#145/#174/#183) is exhausted per project memory; env-side fixes verified working (#197/#198/#199); independent-PPO still plateaus at 15% of optimum

**Comment by rjwalters** [comment]
Starting work on this issue per user request (Phase 1 discriminator: basin trap vs anti-attractor).

**Comment by rjwalters** [comment]
Builder status update:

- PR #278 created with `loom:review-requested`.
- BC pretrain stage **passed**: `gap_closed = 0.934` (>> 0.7 gate) on a 40k-pair/agent, 30-epoch BC fit with the default `PolicyNetwork(hidden_size=64)`. Initial 10k/10-epoch run gave 0.368 (matching #272's data-budget regime); the larger 40k/30ep run shows the architecture itself is capacity-sufficient when the budget is adequate.
- PPO continuation launched on `COMPUTE_HOST_PRIMARY` in tmux `issue270-ppo` (2-way parallel xargs over seeds 42/43/44). Currently severely CPU-starved by parallel #271 sweep on the same host (300+ load avg) — seeds 42/43 are running at ~2% CPU each. Verdict will be added as a PR comment when training completes and `analyze_270.py` is run against the cells.

**Comment by rjwalters** [comment]
## Re-prioritized to TOP per #271 verdict

PR #277 / #271 K=200 verdict ruled out the active-anti-attractor hypothesis (best-of-1000 random nets phase-2 gap_closed ≈ 0.0, regressed to random play). PPO is NOT worse than random search.

The basin-trap reading is now the working thesis. **This issue (#270) is the load-bearing experiment that confirms or further refines it**:

- If BC-init + PPO HOLDS at near-specialist → **basin trap fully confirmed**. The specialist basin is locally stable; PPO just can't reach it from random init.
- If BC-init + PPO COLLAPSES → unexpected third hypothesis: the specialist basin isn't even locally stable. That would be a deeper Goodhart-shaped failure than #271 detected.

Whichever way it resolves, this is the single most important pending experiment. Bumping priority to top of queue.

See `research_notebook/2026-05-17_issue271_best_of_n_verdict.md` for full context.
